### PR TITLE
OOP school library: preserve data

### PR DESCRIPTION
**Pull request by egichuhis**
Hello, please review this PR.

Required:

1. Preserve the data from your app.
2. Load preserved data on startup.
3. On exit from your app save all the data (books, people, and rentals) into JSON files. One per type of data you are storing and named accordingly (e.g. rentals.json)
4. Handle the case when one or more of the files are missing.

### CODE REVIEW

File: savedata.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    puts "Error parsing rentals JSON: #{json_content}"
+    {}
+  end
+
+  def save_people
+    existing_people_data = File.exist?('people.json') ? JSON.parse(File.read('people.json'))['people'] : []
+    new_people_data = @app.people.map(&:to_json)
+    combined_people_data = (existing_people_data + new_people_data).uniq
+    File.write('people.json', JSON.pretty_generate(people: combined_people_data))
+  end
```
Comment (shahadat3669) [inline savedata.rb]: 👏 Good effort on your app! 👏 However, I noticed a minor issue: when exiting by pressing `7`, it seems like the same person is getting duplicated. Let's address this to enhance your code:

- [x] **Duplicate Entry Issue:** It appears that there might be a duplication error in person.json file when exiting with the key `7`. Kindly review your code logic related to exiting and ensure that duplicate entries are handled appropriately. 


https://github.com/egichuhis/school-library-ruby/assets/55840999/0c635dd0-8b83-48d2-9fad-bc6757404ecd



By addressing this duplicate entry concern, you'll ensure a smoother user experience in your app. Keep up the great work! 🌟

File: savedata.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    @app.rentals = rentals_data['rentals']&.map do |rental_data|
+      Rental.from_json(rental_data, @app.books, @app.people)
+    end || []
+
+    @app.rentals.each do |rental|
+      book = Book.find_by_id(rental.book.id, @app.books)
+      book.rentals << rental
+    end
+  end
```
Comment (shahadat3669) [inline savedata.rb:47]: 👏 Good job on incorporating the rental feature! 👏 However, it seems there's a hurdle in running the app again, as an error related to `Error parsing Rental JSON` is occurring. Let's work on resolving this issue to keep the momentum going:

- [x] **Parsing Error Resolution:** It appears that there's a challenge with parsing the Rental JSON. To address this, consider reviewing the code responsible for parsing rental data. Ensure that the JSON format aligns with your parsing logic, and handles potential edge cases or unexpected data gracefully.


image

By updating your code to handle the parsing error, you'll be able to run the app seamlessly after adding rentals. Keep up the good work! 🚀🛠️

File: savedata.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    combined_people_data = (existing_people_data + new_people_data).uniq
+    File.write('people.json', JSON.pretty_generate(people: combined_people_data))
+  end
+
+  def save_books
+    existing_books_data = File.exist?('books.json') ? JSON.parse(File.read('books.json'))['books'] : []
+    new_books_data = @app.books.map(&:to_json)
+    combined_books_data = (existing_books_data + new_books_data).uniq
+    File.write('books.json', JSON.pretty_generate(books: combined_books_data))
+  end
```
Comment (shahadat3669) [inline savedata.rb]: 👏 Kudos on integrating the rental feature! 👏 However, there's a notable concern: after creating a rental and exiting with `7`, a duplicate entry for that book appears in the `books.json` file, restricting you from running the app twice. Let's work on resolving this issue to ensure a seamless experience:

- [x] **Duplicate Entry in books.json:** I observed that exiting with `7` after a rental is causing a duplicate entry in the `books.json` file. To address this, carefully review the section of your code handling exits and ensure it appropriately manages book entries, checking for existing entries before adding a new one.

https://github.com/egichuhis/school-library-ruby/assets/55840999/5a57bf8a-a1e6-4f41-a023-d76990f32e96

By refining this aspect of your code, you'll enable smooth app runs, even after creating a rental. Keep up the excellent work! 🚀📚

**Review by shahadat3669 — CHANGES_REQUESTED** [review]
Hi @egichuhis, @Ochiengsteven, 

Great job so far! 👏 You've done well, and you're almost there for the next project ⏭️, but there are a few issues 🐛 that you still need to work on.




## Don't Worry 🙏

It's okay to have bugs in your code. 🆗 Please see this code review as an opportunity to learn practical coding skills and discover new tips that can improve your code quality 💯 and help you **get even better** 💼.

**I will guide you on how to address these areas down below 👇.**

## Highlights
✅ No linter Errors
✅ Descriptive PR Title
✅ Descriptive PR Description



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

## Have any doubt ❓

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

<table width="100%" align="center">
  <tr>
    <td colspan="4" align="center">
      <strong>Contact me: Shahadat Hossain</strong>
    </td>
  </tr>
  <tr>
    <td align="center">
      <a href="https://github.com/shahadat3669">
        GitHub
      </a>
    </td>
    <td align="center">
      <a href="https://www.linkedin.com/in/shahadat-cseng/">
        LinkedIn
      </a>
    </td>
    <td align="center">
      <a href="https://app.slack.com/client/T47CT8XPG/D05GT4E4WAY/rimeto_profile/U04426SD71V">
        Slack
      </a>
    </td>
    <td align="center">
      <a href="mailto:shahadat3669@gmail.com?subject=Microverse%20Code%20Review">
        Email
      </a>
    </td>
  </tr>
</table>

## ⚠️ WARNING ⚠️

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by egichuhis** [comment]
Thanks @shahadat3669 for noticing the bugs. All fixed and app now works well.

**Review by AmaduKamara — APPROVED** [review]
# Status: Approved 🟢 🟢 🟢

Hi Team, 

Good job working on the changes highlighted by the previous reviewer 👍🏽

Your project is complete!  There is nothing else to say other than... it's time to merge it: ship it:
Congratulations! 🎉 

## Highlights:

✅ You set up Gitflow correctly 👍🏽
✅ The Linter checks passed successfully 👍🏽
✅ App data is preserved in JSON files as required 👍🏽
✅ All features of the app work fine 👍🏽
✅ Good descriptive PR 👍🏽

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

**Pull request merged**